go/tools/gazelle/packages/walk.go

```diff
@@ -126,19 +127,29 @@ func findPackage(c *config.Config, dir string, oldFile *bzl.File, hasTestdata bo
 		rel = ""
 	}
 
-	var goFiles, otherFiles []string
+	// If an existing BUILD file is present, look for comments that exclude
+	// files and rules that produce .go files.
+	var genGoFiles []string
+	var excluded map[string]bool
+	if oldFile == nil {
+		excluded = make(map[string]bool)
+	} else {
+		genGoFiles = findGenGoFiles(oldFile)
+		excluded = findExcludedFiles(oldFile)
+	}
 
 	// List the files in the directory and split into .go files and other files.
 	// We need to process the Go files first to determine which package we'll
 	// generate rules for if there are multiple packages.
+	var goFiles, otherFiles []string
 	files, err := ioutil.ReadDir(dir)
 	if err != nil {
 		log.Print(err)
 		return nil
 	}
 	for _, file := range files {
```

Comment:
I feel like this loop ought to be merged with packages.Walk for a couple of reasons.
First, speed, stating files is expensive.
Second, it would be nice if 
```# gazelle:exclude subdir```
could work to prevent gazelle from descending into a directory, without replicating the parsing an a second place.